engine/crates/engine-v2/src/response_graph/mod.rs

```diff
@@ -0,0 +1,137 @@
+pub struct NodePath(Vec<FieldEdgeId>);
+
+impl NodePath {
+    pub fn empty() -> Self {
+        Self::default()
+    }
+
+    pub fn child(&self, field: FieldEdgeId) -> Self {
+        let mut child = Self(self.0.clone());
+        child.0.push(field);
+        child
+    }
+}
+
+#[derive(Default)]
+pub struct ResponseGraphSchema {
+    fields: Vec<FieldEdge>,
+}
+
+impl ResponseGraphSchema {
+    pub fn push_field(&mut self, field: FieldEdge) -> FieldEdgeId {
+        self.fields.push(field);
+        FieldEdgeId(self.fields.len() - 1)
+    }
+}
+
+impl std::ops::Index<FieldEdgeId> for ResponseGraphSchema {
+    type Output = FieldEdge;
+
+    fn index(&self, index: FieldEdgeId) -> &Self::Output {
+        &self.fields[index.0]
+    }
+}
+
+pub struct ObjectNodeId(usize);
+
+pub struct ResponseGraph {
+    schema: ResponseGraphSchema,
+    root: ObjectNodeId,
+    objects: Vec<ObjectNode>,
+}
+
+// Only supporting additions for the current graph. Deletion are... tricky
+// It shouldn't be that difficult to know whether a remaining plan still needs a field after
+// execution plan creation. But it's definitely not efficient currently. I think we can at
+// least wait until we face actual problems. We're focused on OLTP workloads, so might never
+// happen.
+impl ResponseGraph {
+    pub fn new(schema: ResponseGraphSchema) -> Self {
+        Self {
+            schema,
+            root: ObjectNodeId(0),
+            objects: vec![ObjectNode { fields: vec![] }],
+        }
+    }
+
+    pub fn push_object(&mut self, object: ObjectNode) -> ObjectNodeId {
+        self.objects.push(object);
+        ObjectNodeId(self.objects.len() - 1)
+    }
+
+    pub fn node_ids(&self, path: &NodePath) -> Vec<ObjectNodeId> {
+        todo!()
+    }
+
+    // Using schema field names for federation should work. But we'll probably
+    // want a HashMap<FieldId, String> for connectors to rename fields however they need it to.
+    // For inputs the selection_set contains one edge per FieldId (probably? surely?)
+    pub fn nodes_as_json_without_alias(
+        &self,
+        nodes: &[ObjectNodeId],
+        selection_set: &NodeSelectionSet,
+    ) -> Vec<serde_json::Value> {
+        todo!()
+    }
+}
+
+impl std::ops::Index<ObjectNodeId> for ResponseGraph {
+    type Output = ObjectNode;
+
+    fn index(&self, index: ObjectNodeId) -> &Self::Output {
+        &self.objects[index.0]
+    }
+}
+
+pub struct ObjectNode {
+    fields: Vec<(FieldEdgeId, Node)>,
+}
+
+pub enum Node {
+    // and other stuff
+    String(String),
+    List(Vec<Node>),
+    Object(ObjectNodeId),
+}
+
+// Maybe an enum? Need to support `__typename`
+pub struct FieldEdge {
+    // probably needs a better name. it's the position for requested fields. For added fields,
+    // it's the position of the query field that needed it.
+    pub pos: Pos,
+    // resolving fragments eagerly, it makes manipulating SelectionSet easier during planning.
+    pub type_condition: Option<TypeCondition>,
```

Comment:
Have you given any thought to the other uses of fragments?  I can see this approach working alright for type conditions, but not so sure about `@defer`, `@skip`, `@include`?  Not so sure

Reply:
excellent point. for `@skip` & `@include` it shouldn't be hard as the condition can be detected quite easily. For `@defer`, it's definitely more tricky. I think in this case we would probably have a `deferred` field in the `NodeSelectionSet` containing the deferred fragment fields. A data source might only be responsible of only some parts of fragment fields, so I'm a bit skeptical of keeping the fragment notion which does add more complexity. The only benefit I see is more readable queries generated against subgraphs. A valid one, but not critical imho.